lib/eventFactory.js

```diff
@@ -2,9 +2,131 @@
+ * @param  {Number}   eventId                    Event id
+ */
+function createBuilds(pipeline, config, eventId) {
+    const startFrom = config.startFrom;
+    let jobsToStart = [];
+
+    // If no startFrom, then do nothing. Remove once we switch to new workflow design
+    if (!startFrom) {
+        return null;
+    }
+
+    /* eslint-disable global-require */
+    const BuildFactory = require('./buildFactory');
+    const buildFactory = BuildFactory.getInstance();
+    const JobFactory = require('./jobFactory');
+    const jobFactory = JobFactory.getInstance();
+    /* eslint-enable global-require */
+
+    return pipeline.jobs.then((jobs) => {
+        let nextJobs;
+
+        // When startFrom is ~commit
+        if (startFrom === '~commit') {
+            return pipeline.sync()
+            .then((p) => {
+                nextJobs = WorkflowParser.getNextJobs(p.workflowGraph, { trigger: startFrom });
+                jobsToStart = jobs.filter(j => nextJobs.includes(j.name) && j.state === 'ENABLED');
+
+                return Promise.resolve([]);
+            });
+        }
+
+        // When startFrom is jobName
+        if (startFrom !== '~pr') {
+            jobsToStart = jobs.filter(j => j.name === startFrom && j.state === 'ENABLED');
+
+            return Promise.resolve([]);
+        }
+
+        // When startFrom is ~pr
+        return pipeline.syncPR(config.prNum)
+        .then(() => {
+            // Get next jobs for when startFrom is ~pr
+            nextJobs = WorkflowParser.getNextJobs(pipeline.workflowGraph, {
+                trigger: startFrom,
+                prNum: config.prNum
+            });
+
+            // To start all existing and ENABLED PR-prNum jobs
+            const PRJobArray = jobs.filter(j => j.name.startsWith(`PR-${config.prNum}:`));
+            const enabledPRJobArray = PRJobArray.filter(j => j.state === 'ENABLED');
+
+            jobsToStart = enabledPRJobArray;
+
+            // Get all the missing PR- job names
+            const existingPRJobNames = PRJobArray.map(p => p.name);
+            const missingPRJobNames = nextJobs.filter(j => !existingPRJobNames.includes(j));
+            // Get the job name part, e.g. main from PR-1:main
+            const missingJobNames = missingPRJobNames.map(name => name.split(':')[1]);
+            const missingJobs = jobs.filter(j => missingJobNames.includes(j.name));
+
+            return pipeline.getConfiguration(config.prRef)
+            .then((parsedConfig) => {
+                console.log('parsedConfig: ', parsedConfig.jobs);
+
+                // Create missing PR jobs
+                return Promise.all(missingJobs.map(j =>
+                    createJob(jobFactory, {
+                        permutations: parsedConfig.jobs[j.name],
+                        jobName: `PR-${config.prNum}:${j.name}`,
+                        pipelineId: pipeline.id
+                    })
+                ));
+            });
+        });
+    })
+    .then((newJobs) => {
```

Comment:
can you modify logic to take `jobsToStart` here ? Then you don't need the global variable `jobsToStart`, and no need to do the `concat` business inside this function, which is purely a `PR` job thingy